spec/unit/matrixrtc/mocks.ts

```diff
@@ -18,27 +18,31 @@ import { EventType, MatrixEvent, Room } from "../../../src";
 import { CallMembershipData } from "../../../src/matrixrtc/CallMembership";
 import { randomString } from "../../../src/randomstring";
 
-export function makeMockRoom(
-    memberships: CallMembershipData[],
-    getLocalAge: (() => number) | undefined = undefined,
-): Room {
+export function makeMockRoom(memberships: CallMembershipData[], localAge: number | null = null): Room {
     const roomId = randomString(8);
+    // If makeMockRoom is used instead of one of the other functions in this file,
+    // we store it so that subsequent calls calculate the expiration based on it using
+    // Date.now(). So, jest.advanceTimersByTime() can be used to simulate the passage of time.
+
+    // The actual time of the underlying mocked events will be: localTimestamp ?? localAge ?? 0
+    const localTimestamp = Date.now() - (localAge ?? 10);
```

Comment:
isn't this exactly what happens in `makeMockRoomState` if you just pass in a `localAge`? Can we omit this line? (And maybe move the comment above to `mockRTCEvent`)

Reply:
This is the core part of it. We need to cache the `localTimestamp` so that subsequent calls to `getLiveTimeline` (which then calls: `makeMockRoomState`) use this `localTimestamp` instead of the `localAge` (since `localAge` is ignored if `localTimestamp` is the present as described in the comment.)

Otherwise we would call `Date.now() - localAge` at the time where we call `getLiveTimeline` so we could of course not use the initial age but need to update it with the elapsed time. But when calling `getLiveTimeline` we cannot access the age anymore since age is not a method anymore.

By cahing the localTimestamp here we only use the timestamp when calling  `getLiveTimeline` and compute the age based on that so it is all in sync.

Reply:
> `makeMockRoomState` can be called indirectly in the tested code when the mocked `getLiveTimeline` is called. If we mock it with a `localAge` than the timestamp would be recomputed using `Date.now()` - `localAge`. So we would recompute the local timestamp "accidently" while calling `getLiveTimeline` which breaks a lot of tests (It would be a bug). Caching it here and then alsways using `localTimestamp` allows to always recompute the local age based on the timestamp. So both are correct.

I'm afraid I'm not following you. What do you mean by "if we mock it" - mock *what*?

Currently your code is:

```typescript
    const localTimestamp = Date.now() - (localAge ?? 10);
    return {
        roomId: roomId,
        getLiveTimeline: jest.fn().mockReturnValue({
            getState: jest.fn().mockReturnValue(makeMockRoomState(memberships, roomId, localAge, localTimestamp)),
        }),
    } as unknown as Room;
```

... which is exactly equivalent to:

```typescript
    const localTimestamp = Date.now() - (localAge ?? 10);
    const mockedRoomState = makeMockRoomState(memberships, roomId, localAge, localTimestamp);
    return {
        roomId: roomId,
        getLiveTimeline: jest.fn().mockReturnValue({
            getState: jest.fn().mockReturnValue(mockedRoomState),
        }),
    } as unknown as Room;
```

... which again is equivalent to

```typescript
    const mockedRoomState = makeMockRoomState(memberships, roomId, localAge,  Date.now() - (localAge ?? 10);
    return {
        roomId: roomId,
        getLiveTimeline: jest.fn().mockReturnValue({
            getState: jest.fn().mockReturnValue(mockedRoomState),
        }),
    } as unknown as Room;
```

... and I don't understand why specifying an explicit value for `makeMockRoomState`'s fourth argument is any different to letting it use its default.

Reply:
> ... which is exactly equivalent to:

It is not equivalent since the initial version will call makeMockRoomState again while your proposed one wont.
This is exactly what I treid to achieve with the localTimestamp prop. Because when calling `makeMockRoomState` (like it initially did) localTimestamp gets recomputed with `Date.now() - (localAge ?? 10)` using the current `Date.now()`.

But just storing the whole mockedRoomState is muuch easier to read!
See this change: https://github.com/matrix-org/matrix-js-sdk/pull/3900/commits/54b71bb729090a7534a54bd6dbfdc3e3d738d66a

(I think this might be exactly what you tried to communicate. Thanks. I removed a couple of comments since the things the comments addressed are quiet obvious now.)

Reply:
For the record: it turns out that `makeMockRoomState` didn't immediately call `mockRTCEvent`, but instead delayed it until `RoomState.getStateEvents` was called. Hence, passing a `localAge` into `makeMockRoomState` would produce "unintuitive" effects. Sorry, I should have spotted this during earlier review.

`makeMockRoomState` is updated as of 42be793a5 so this is resolved.